pkg/messaging/broker.go

```diff
@@ -0,0 +1,168 @@
+package messaging
+
+import (
+	"time"
+
+	"github.com/cskr/pubsub"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/openservicemesh/osm/pkg/announcements"
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha1"
+	"github.com/openservicemesh/osm/pkg/k8s/events"
+	"github.com/openservicemesh/osm/pkg/metricsstore"
+)
+
+// NewBroker returns a new message broker instance and starts the internal goroutine
+// to process events added to the workqueue.
+func NewBroker(stopCh <-chan struct{}) *Broker {
+	b := &Broker{
+		queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		proxyUpdatePubSub: pubsub.New(0),
+		kubeEventPubSub:   pubsub.New(0),
+		certPubSub:        pubsub.New(0),
+	}
+
+	go b.run(stopCh)
+
+	return b
+}
+
+// GetProxyUpdatePubSub returns the PubSub instance corresponding to proxy update events
+func (b *Broker) GetProxyUpdatePubSub() *pubsub.PubSub {
+	return b.proxyUpdatePubSub
+}
+
+// GetKubeEventPubSub returns the PubSub instance corresponding to k8s events
+func (b *Broker) GetKubeEventPubSub() *pubsub.PubSub {
+	return b.kubeEventPubSub
+}
+
+// GetCertPubSub returns the PubSub instance corresponding to certificate events
+func (b *Broker) GetCertPubSub() *pubsub.PubSub {
+	return b.certPubSub
+}
+
+// run starts a goroutine to process events from the workqueue until
+// signalled to stop on the given channel.
+func (b *Broker) run(stopCh <-chan struct{}) {
+	// Start the goroutine workqueue to process kubernetes events
+	// The continuous processing of items in the workqueue will run
+	// until signalled to stop.
+	// The 'wait.Until' helper is used here to ensure the processing
+	// of items in the workqueue continues until signalled to stop, even
+	// if 'processNextItems()' returns false.
+	go wait.Until(
+		func() {
+			for b.processNextItem() {
+			}
+		},
+		time.Second,
+		stopCh,
+	)
+}
+
+// processEvent processes an event dispatched from the workqueue.
+// It does the following:
+// 1. If the event must update a proxy, it publishes a proxy update message
+// 2. Processes other internal control plane events
+// 3. Updates metrics associated with the event
+func (b *Broker) processEvent(msg events.PubSubMessage) {
+	// Update proxies if applicable
+	if shouldUpdateProxy(msg) {
+		b.proxyUpdatePubSub.Pub(msg, announcements.ProxyUpdate.String())
+		metricsstore.DefaultMetricsStore.ProxyBroadcastEventCount.Inc()
+	}
+
+	// Publish event to other interested clients, e.g. log level changes, debug server on/off etc.
+	b.kubeEventPubSub.Pub(msg, msg.Kind.String())
```

Comment:
Based on the second point [here](https://pkg.go.dev/k8s.io/client-go/util/workqueue#pkg-overview):

> if an item is added multiple times before it can be processed, it will only be processed once.

Is it possible that some kubernetes events will not be propagated? I can see where that wouldn't matter in most cases, but could that cause issues in certain scenarios, like if we're counting events?

Reply:
I haven't seen an issue with this, and I am unsure if the link you shared applies here. We do not drop any events as such, the rate limiter could if the same object is repeatedly queued, which does not happen right now. I don't see an issue with the event counters either.